File: js/sort/merge.js
/**
 * 合并排序
 */

var mergeSort = function(arr) {
  function merge(leftArr, rightArr) {
    var mergeResult = []
    while(leftArr.length && rightArr.length) {
      if(leftArr[0] < rightArr[0]) {
        mergeResult.push(leftArr[0])
        leftArr.shift()
      } else {
        mergeResult.push(rightArr[0])
        rightArr.shift()
      }
    }

    return mergeResult.concat(leftArr.concat(rightArr))
  }

  if(arr.length<2)
    return

  var midIndex = parseInt((arr.length-1) / 2)
  var left = arr.slice(0, midIndex)
  var right = arr.slice(midIndex)
  console.log(left,right)

  return merge(mergeSort(left), mergeSort(right))
}

console.log(mergeSort([3,2,5]))

module.exports = mergeSort

File: python/sort/merge.py
def mergeSort(arr):
    def merge(left, right):
        result = []
        while(len(left) != 0 and len(right) != 0):
            if left[0] > right[0]:
                result.append(right[0])
                right.pop(0)
            else:
                result.append(left[0])
                left.pop(0)
        if len(left) != 0:
            result[len(result):] = left
        if len(right) != 0:
            result[len(result):] = right
        return result

    length = len(arr)
    if length < 2:
        return arr

    midIndex = int(length / 2)
    leftArr = arr[0:midIndex]
    rightArr = arr[midIndex:]

    return merge(mergeSort(leftArr), mergeSort(rightArr))
